### PR TITLE
Removes -2 special case and specialization from pointwise apply

### PR DESCRIPTION
**Pull request by mruberry**
Currently in pointwise apply there are four special cases:

-2, a 1D tensor with stride == 1
-1, a nD tensor
1, a 1D tensor with stride != 1
2, a 2D (noncontiguous) tensor

There are also template specializations for the -2 special case in THC's OffsetInfo and and ATen's IndexToOffset. 

The practical difference, however, between the -2 and 1 code paths is only a multiplication. The -2 code path simply returns its input (implicitly * 1) while the 1 code path returns input * stride. Since the pointwise apply kernels are expected to be bandwidth bound, performing this additional multiplication likely has no performance impact. To demonstrate this empirically I reviewed the THC pointwise apply 1 and 2 kernels on 1D tensors of 3 billion, 1 billion, half a billion, and 1 million floats. Operations were performed on these tensors at strides of 1, 2, and 4. There was no significant performance difference between Master and this PR, with maximum variation of only .3% between the two, well within the observed noise of the experiments I was running (avg of 10 reps for the tensors of 1 billion or greater size, and 100 reps for the smaller tensors).

While there is (almost certainly) no performance change from this fix, it does reduce build time and simplify the existing code. Pointwise apply makes note of its sensitivity to excess template instantiations, after all.  

Note that, for readability, this PR currently leaves the dims = 1 instantiations of OffsetInfo and IndexToOffset. Without this the code path would be the same, although the reader would have to review the code somewhat carefully to determine this. Given the importance of 1D tensors it seemed appropriate to leave these specializations as a guide to readers.

### CODE REVIEW

**Comment by apaszke** [comment]
@mruberry Do you have any benchmarks showing that it really doesn't matter?

**Review by apaszke — APPROVED** [review]
Thanks, LGTM! Can we please remove the specializations too? I know they might improve readability for now, but ultimately that's more code to maintain, and it's code that should always stay in sync with the generic version. It shouldn't be too hard to review anyway.

File: aten/src/THC/THCTensorInfo.cuh [hunk 1]
```diff
@@ -302,7 +299,6 @@ struct OffsetInfo {
 };
 
 // For 1D tensors the offset equals linear index * stride.
-// Note: this specialization for readability only
 template <typename T, typename IndexType>
```
Comment (apaszke) [inline aten/src/THC/THCTensorInfo.cuh]: This could probably be removed as well

**Comment by mruberry** [comment]
Always happy to remove code ;)

I aligned the IndexToOffset code (for some reason there were 3 slightly different versions) and removed the IndexToOffset specialization in TensorInfo.cuh (under ATen). I was mistaken that the OffsetInfo specialization would be so easy to remove, however, since it allocates arrays, and the 1D case does not need to do so. I've kept that one for now.

I am currently working on a follow-up PR, however, that I expect will improve performance and further adjust the THCTensorInfo.cuh code in a way that will render this issue moot.

**Review by apaszke — APPROVED** [review]

File: aten/src/THC/THCTensorInfo.cuh (hunk 1 above)
Comment (mruberry) [reply aten/src/THC/THCTensorInfo.cuh]: Wait for my next PR and whp it will be.

**Pull request merged**